datafusion/optimizer/src/simplify_expressions/function_simplifier.rs

```diff
@@ -0,0 +1,54 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module implements a rule that do function simplification.
+
+use datafusion_common::tree_node::TreeNodeRewriter;
+use datafusion_common::Result;
+use datafusion_expr::simplify::Simplified;
+use datafusion_expr::simplify::SimplifyInfo;
+use datafusion_expr::{expr::ScalarFunction, Expr, ScalarFunctionDefinition};
+
+pub(super) struct FunctionSimplifier<'a, S> {
```

Comment:
I think this would be more discoverable (and avoid another tree walk) if the code was inlined into the existing ExprSimplifier (the giant `match` statement that handles all the different kinds of `Expr`s) -- perhaps right after the other rewrites for `Expr::ScalarFunction` ?